### PR TITLE
Support sharded target logits for EAGLE3 online training

### PR DESCRIPTION
**Pull request by Yukino256**
## Motivation

This PR adds a memory-efficient target-logits path for EAGLE3 online training with the SGLang target backend.

Previously, online EAGLE3 training materialized full target logits on every TP rank after SGLang's tensor-parallel logits all-gather. For long-context training, this creates large redundant logits tensors and can lead to OOM. This becomes especially problematic when enabling sequence parallel training, where the draft model only needs the local sequence shard instead of full-sequence logits on every rank.

## Modifications

- Add `--shard-target-logits` to `scripts/train_eagle3.py`.
- For SGLang target models, optionally disable SGLang's target-logits TP all-gather and redistribute logits explicitly.
- Add two sharded target-logits redistribution paths:
  - Non-SP online training: redistribute from full-batch/local-vocab logits to local-batch/full-vocab logits.
  - USP online training: redistribute from full-sequence/local-vocab logits to local-sequence/full-vocab logits.
- Support online USP training with `tp_size >= sp_size` when `tp_size % sp_size == 0`.
- Align target hidden states, input IDs, attention masks, loss masks, and position IDs with the local USP sequence shard.
- Keep the original behavior unchanged unless `--shard-target-logits` is enabled.

## Related Issues

N/A

## Accuracy Test

Tested EAGLE3 online training on DeepSeek-V3 with the SGLang target backend.

Configuration:

```bash
--target-model-backend sglang
--tp-size 8
--sp-ring-size 1
--sp-ulysses-size 8
--attention-backend usp
--max-length 32768
--sglang-mem-fraction-static 0.8
--shard-target-logits
```

The 32K training run proceeds without OOM. Training loss decreases normally and training accuracy increases normally.

Training loss / accuracy curves:

image

Downstream acceptance length on the related evaluation data: **2.53**. This is roughly consistent with the previous non-SP training result we used.

SGLang Inference/evaluation Logs:

image

## Benchmark & Profiling

This PR is primarily intended to reduce target-logits memory usage during EAGLE3 online training.

Observed behavior:

- Before this change, target logits were materialized redundantly on every TP rank.
- With `--shard-target-logits`, each rank only keeps the target logits needed by its local batch shard or local USP sequence shard.
- In the tested DeepSeek-V3 32K online USP setup, training runs successfully with `--sglang-mem-fraction-static 0.8`.

No throughput benchmark is included yet.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

File: specforge/modeling/target/eagle3_target_model.py [hunk 1]
```diff
@@ -688,93 +814,119 @@ def generate_eagle3_data(
...
+                    ring_chunk = usp_chunk_size * sp_ulysses_size
+                    ring_start = sp_ring_rank * ring_chunk
+                    kept_position_ids.append(
+                        torch.arange(
+                            ring_start,
+                            ring_start + ring_chunk,
+                            dtype=torch.long,
+                            device=input_id.device,
+                        ).unsqueeze(0)
+                    )
```
[the remark this answers is not in the document]
Comment (Yukino256) [reply specforge/modeling/target/eagle3_target_model.py:896]: This is intentional and follows the existing offline USP preprocessing logic. In USP mode, position_ids are not local input slice positions. They need to match the Ulysses all-to-all expanded sequence length consumed by UspAdapter, i.e. (local_len - ttt_length) * sp_ulysses_size. The online path mirrors process_data_usp so online and offline USP use consistent position ids.

File: specforge/modeling/target/sglang_backend/utils.py [hunk 2]
```diff
@@ -25,6 +26,222 @@ class ReplacedLogitsProcessorEagle3Output:
...
+            (num_samples, local_len, local_vocab_size),
+            dtype=logits.dtype,
+            device=logits.device,
+        )
+        for sample_idx, seq_len in enumerate(input_lens):
+            end = min(start + local_len, seq_len)
+            valid_len = max(0, end - start)
+            if valid_len > 0:
+                send[sample_idx, :valid_len] = padded_logits[sample_idx, start:end]
+        input_list.append(send)
```
[the remark this answers is not in the document]
Comment (Yukino256) [reply specforge/modeling/target/sglang_backend/utils.py]: Good point. The current implementation favors simplicity and correctness, but it does allocate one send tensor per TP rank before all_to_all. This can increase peak memory and allocator fragmentation for long-context runs. I will update it to pack the send tensors into a single contiguous buffer and use views for all_to_all inputs. This should reduce allocation overhead and make the path more stable, while keeping the same redistribution semantics.

File: specforge/modeling/target/sglang_backend/utils.py [hunk 3]
```diff
@@ -25,6 +26,222 @@ class ReplacedLogitsProcessorEagle3Output:
...
+            (local_len, hidden_states.shape[1]),
+            dtype=hidden_states.dtype,
+            device=hidden_states.device,
+        )
+        end = min(start + local_len, seq_len)
+        valid_len = max(0, end - start)
+        if valid_len > 0:
+            sample_start = sample_offsets[sample_idx]
+            local[:valid_len] = hidden_states[sample_start + start : sample_start + end]
+        local_tokens.append(local)
```
[the remark this answers is not in the document]
Comment (Yukino256) [reply specforge/modeling/target/sglang_backend/utils.py]: Good catch. I agree this can cause unnecessary allocation overhead. I will change it to pre-allocate a contiguous output buffer for all samples and fill the per-sample shards in place.

**Comment by Yukino256** [comment]
Hi @jiapingW  could you please help take a look?

This PR solves a similar issue as #524, but extends it to the `online + USP + sharded target logits` setting. It avoids redundant target logits across TP ranks and makes the SGLang target logits align with the local USP sequence shard.

I tested it on DeepSeek-V3 32K online training (`tp=8`, `sp=8`, `mem_fraction_static=0.8`). Training runs without OOM, loss/accuracy look normal, and the downstream acceptance length is 2.53, roughly consistent with our previous non-SP result.

Thanks!

**Comment by jiapingW** [comment]
Great. I’ll review it soon.

**Comment by jiapingW** [comment]
@Yukino256 Hi, I apologize for the late reply. We recently refactored the entire online training logic; currently, we're using  SGL server for separate training and inversion online training. Could you please modify the current PR based on the current `main` branch? Thank you very much for your contribution.
